File: c/switch.c
#define G_DISABLE_ASSERT 1
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "common.h"

#define C   r[ins & 7]
#define B   r[(ins >> 3) & 7]
#define A   r[(ins >> 6) & 7]
#define A2  r[(ins >> 25) & 7]

int main(int ac, const char *av[]) {
    pint r[8] = {0};
    pint pc = 0 ;

    if (ac != 2) DIE("usage: %s <program.um>", *av) ;

    load_file(av[1]) ;

    for (;;) {
        g_assert(pc < zero->ncells) ;
        pint ins = zero->cells[pc] ;

        switch ((ins >> 28) & 15) {
            case  0: if (C) A = B ; break ;
            case  1: A = array_get(B, C) ; break ;
            case  2: array_set(A, B, C) ; break ;
            case  3: A = (uint32_t)(B + C) ; break ;
            case  4: A = (uint32_t)(B * C) ; break ;
            case  5: A = ((uint32_t)B) / ((uint32_t)C) ; break ;
            case  6: A = (uint32_t)~(B & C) ; break ;
            case  7: g_message("HALT") ; exit(0) ; break ;
            case  8: B = array_alloc(C) ; break ;
            case  9: array_free(C) ; break ;
            case 10: fputc((char)C, stdout) ; break ;
            case 11: C = vm_fgetc() ; break ;
            case 12: {
                         if (B) {
                            array_free((pint)zero) ;
                            struct array *b = (struct array *)B ;
                            zero = (struct array *)array_alloc(b->ncells) ;
                            memcpy(zero->cells, b->cells,
                                    b->ncells * sizeof(pint)) ;
                         }
                         pc = C ;
                         continue ;
                     }
            case 13: A2 = ins & 0x1ffffff ; break ;
            default:
                     DIE("unknown instruction %lx", ins) ;
        }
        pc ++ ;
    }

    return 0 ;
}

File: nim/switch.nim
import os, streams, sequtils, posix

type
  Pint = uint32
  Array = seq[Pint]

proc die(s: string) =
  echo s
  quit -1

var arrays: seq[ref[Array]] = @[]
var free: Array = @[]

proc allocArray(size: int): Pint =
  let x = new Array
  x[] = repeat(Pint 0, size)

  if free.len == 0:
    arrays &= x
    return Pint arrays.high

  let a = free.pop
  assert arrays[int a] == nil
  arrays[int a] = x
  return a

proc freeArray(a: Pint) =
  free &= a
  arrays[int a] = nil

proc loadFile(fname: string): Pint =
  let fs = int getFileSize(fname)
  
  if fs mod sizeof(Pint) != 0:
    die "weird file size: " & $fs

  let a = allocArray(fs div sizeof(Pint))
  if newFileStream(fname).readData(addr arrays[int a][0], fs) != fs:
    die "short read"

  for p in arrays[int a][].mitems: p = ntohl p
  return a

proc main() =
  let argv = commandLineParams()
  if argv.len != 1:
    die "please specify the UM program"

  doAssert loadFile(argv[0]) == 0
  var pc = Pint 0
  var r: array[Pint 8, Pint]
  var zero = arrays[0]

  template C(): untyped = r[ins and 7]
  template B(): untyped = r[(ins shr 3) and 7]
  template A(): untyped = r[(ins shr 6) and 7]
  template A2(): untyped = r[(ins shr 25) and 7]

  while true:
    {.computedGoto.}
    var ins = zero[int pc]

    case (ins shr 28) and 15:
      of 0:
        if C != 0: A = B
      of 1:
        A = arrays[int B][int C]
      of 2:
        arrays[int A][int B] = C
      of 3:
        A = B + C
      of 4:
        A = B * C
      of 5:
        A = B div C
      of 6:
        A = not(B and C)
      of 7:
        echo "HALT"
        return
      of 8:
        B = allocArray(int C)
      of 9:
        freeArray C
      of 10:
        doAssert stdout.writeBytes([uint8 C], 0, 1) == 1
      of 11:
        var b = [uint8 0]
        let read = stdin.readBytes(b, 0, 1)

        C =
          if read == 0:
            not Pint 0
          else:
            b[0]
      of 12:
        if B != 0:
          arrays[0][] = arrays[int B][]
          zero = arrays[0]
        pc = C - 1
      of 13:
        A2 = Pint(ins and 0x1ffffff)
      of 14, 15:
        die "unknown opcode"

    inc pc

main()
